### PR TITLE
Making Linkage Checker plugin more robust

### PR DESCRIPTION
**Pull request by suztomo**
Rather than failing completely upon unexpected values/conditions, this PR shows the partial results.

Screen Shot 2021-08-11 at 17 09 31

### CODE REVIEW

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassPathResultTest.java [hunk 1]
```diff
@@ -115,21 +114,6 @@ public void testFormatDependencyPaths_threePathsForA() {
...
-  public void testFormatDependencyPaths_irrelevantJar() {
-    AnnotatedClassPath annotatedClassPath =
-        AnnotatedClassPath.fromMultimap(ImmutableListMultimap.of(jarA, dependencyPath_A));
-
-    ClassPathResult classPathResult = new ClassPathResult(annotatedClassPath, ImmutableSet.of());
-
-    try {
-      classPathResult.formatDependencyPaths(ImmutableList.of(jarB));
-      fail("The irrelevant JAR file should be invalidated.");
-    } catch (IllegalArgumentException expected) {
```
Comment (suztomo) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassPathResultTest.java:128]: With this PR, the method doesn't throw IllegalArgumentException any more.

Comment (elefeint) [reply dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassPathResultTest.java:128]: We have examples of capturing output, if you still want to test that the warning was printed: https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/366b3d9530efda9b9b1e7dcd92d0284a38034c25/spring-cloud-gcp-samples/spring-cloud-gcp-integration-pubsub-sample/spring-cloud-gcp-integration-pubsub-sample-receiver/src/test/java/com/example/ReceiverIntegrationTest.java#L68-L71

File: gradle-plugin/src/main/java/com/google/cloud/tools/dependencies/gradle/LinkageCheckTask.java [hunk 2]
```diff
@@ -217,16 +217,27 @@ private void recordDependencyPaths(
...
-        recordDependencyPaths(output, stack, targetCoordinates);
+
+        if (stack.contains(child)) {
+          // Circular dependency check
+          getLogger()
+              .error(
+                  "Circular dependency for: "
+                      + resolvedDependencyResult
+                      + "\n The stack is: "
+                      + ImmutableList.copyOf(stack));
```
Comment (elefeint) [inline gradle-plugin/src/main/java/com/google/cloud/tools/dependencies/gradle/LinkageCheckTask.java:228]: Probably don't need the defensive copy here.

**Review by elefeint — APPROVED** [review]

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassPathResultTest.java (hunk 1 above)
Comment (suztomo) [reply dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassPathResultTest.java:128]: The warning is not critical message. I keep as it is.

File: gradle-plugin/src/main/java/com/google/cloud/tools/dependencies/gradle/LinkageCheckTask.java (hunk 2 above)
Comment (suztomo) [reply gradle-plugin/src/main/java/com/google/cloud/tools/dependencies/gradle/LinkageCheckTask.java:228]: That's right. Removed.

**Pull request merged**